teps/0091-trusted-resources.md

```diff
@@ -0,0 +1,335 @@
+enterprises, for example.
+
+## Requirements
+
+<!--
+Describe constraints on the solution that must be met. Examples might include
+performance characteristics that must be met, specific edge cases that must
+be handled, or user scenarios that will be affected and must be accomodated.
+-->
+
+- There is a process by which a Tekton Resource can be marked as trusted meaning the Resource has been explicitly signed off by an authenticated third party.
+- Verification can ensure the contents of the Resource has not been modified since
+  being marked as trusted.
+- It is not possible to modify the marked Resource in such a way that it will still
+  pass verification.
+- Verification can be performed by Tekton Pipelines; Tekton Chains can see
+  whether or not the verification occurred and if it was successful.
+- A Pipeline Author can indicate that a Task must be verified in order
+  for Chains to create a provenance attestation.
+- A Pipeline Author can configure a run to fail if a Resource cannot be verified.
+- Ideally, the solution should make it clear how to obtain the public key for community-provided, trusted resources.
+
+## Proposal
+
+Sigstore Cosign (https://github.com/sigstore/cosign) has mechanisms to securely
+sign a given OCI image or other artifacts including binaries, scripts etc. This provides a solid footing to leverage Cosign to verify Tekton Resources.
+
+This proposal will introduce a new admission webhook to Tekton to verify a
+Tekton Resource and can indicate that the verification must occur.
+
+If a Resource fails verification, the corresponding TaskRun will be
+annotated as such. The Pipeline Author will have optional configuration to stop
+and fail the run on verification failure.
+
+**IMPORTANT:** A Task refers to other images. For the initial delivery of this
+TEP, all static image references within a Task fetched as a Remote Resource
+are suggested to be referenced by digest
+(i.e., `...@sha256:abcdef`) in order for the verification to succeed.
+In later iterations, this can be extended such that tag-referenced images can
+_themselves_ be verified for the overall Task to pass verification. However,
+this will be left out of the scope of this TEP.
+
+The Pipeline Author will be able to provide optional configuration to Tekton
+Chains such that Chains will *only* consider results from Tasks that were
+verifiable (and, following the previous rule, were successfully verified).
+
+
+### Risks and Mitigations
+
+* Mutating webhook is a risk of the proposal. The proposed verification will happen after the mutating webhook, so the content of a Resource may be mutated and fail the verification.
+
+  Some possible  mitigations include the following:
+
+  1. Update the Resources to add the missing mutating values (e.g. Use `SetDefaults`) everytime we bump the version of Tekton Pipeline.
+  2. Avoid mutating Tekton Resources or move the mutating after verification.
```

Comment:
Wouldn't another mitigation be moving the verification from the webhook to the reconciler?

Reply:
I can do some research this week and see if there are better ways!

Reply:
You are right. One option is that we apply the SetDefaults() each time on reconcile instead of at admission time (so the object stored in etcd is not mutated)

Another possibility might be having the CLI that is doing the signing make a dry-run call to the API server to get the  resource defaults set and sign that https://kubernetes.io/docs/reference/using-api/api-concepts/#dry-run

Reply:
Oh thanks!
The dryrun is really helpful, Billy just also mentioned this.
Right now it is similar that I add call setdefaults during signing, the benefit is that it will not depend a running cluster, but it cannot guarantee that the setdefault is the same as the cluster's tekton.  So I think what you're suggesting is better!

Reply:
> +1 for moving defaults to the reconciler - the less we mutate the config before verification the better.

I think people may not like this idea but will keep this as an option.  It is kind of against the original desgin of admission webhooks? All the request objects need to be mutated and then validate, so some validating logic assume that this object must be mutated.
If we move mutating into reconciler then what stored in etcd is not mutated and fully validated.  

If we can do verification->mutating->verification then it may be better?